csharp/rust/src/lib.rs

```diff
@@ -1,284 +1,472 @@
+    InitResult {
+        success: true as c_int,
+        logger_level: match logger_level {
+            logger_core::Level::Error => ELoggerLevel::Error,
+            logger_core::Level::Warn => ELoggerLevel::Warn,
+            logger_core::Level::Info => ELoggerLevel::Info,
+            logger_core::Level::Debug => ELoggerLevel::Debug,
+            logger_core::Level::Trace => ELoggerLevel::Trace,
+            logger_core::Level::Off => ELoggerLevel::Off,
+        },
     }
-    result
 }
 
-fn convert_vec_to_pointer<T>(mut vec: Vec<T>) -> (*const T, usize) {
-    vec.shrink_to_fit();
-    let vec_ptr = vec.as_ptr();
-    let len = vec.len();
-    // TODO use `Box::into_raw`
-    std::mem::forget(vec);
-    (vec_ptr, len)
-}
-
-/// Execute a command.
-/// Expects that arguments will be kept valid until the callback is called.
+/// # Summary
+/// Creates a new client to the given address.
+///
+/// # Input Safety (in_...)
+/// The data passed in is considered "caller responsibility".
+/// Any pointers hence will be left unreleased after leaving.
+///
+/// # Output Safety (out_... / return ...)
+/// The data returned is considered "caller responsibility".
+/// The caller must release any non-null pointers.
 ///
-/// # Safety
-/// * `client_ptr` must not be `null`.
-/// * `client_ptr` must be able to be safely casted to a valid [`Box<Client>`] via [`Box::from_raw`]. See the safety documentation of [`std::boxed::Box::from_raw`].
-/// * This function should only be called should with a pointer created by [`create_client`], before [`close_client`] was called with the pointer.
-/// * `key` and `value` must not be `null`.
-/// * `key` and `value` must be able to be safely casted to a valid [`CStr`] via [`CStr::from_ptr`]. See the safety documentation of [`std::ffi::CStr::from_ptr`].
-/// * `key` and `value` must be kept valid until the callback is called.
-/// * `route_info` could be `null`, but if it is not `null`, it must be a valid [`RouteInfo`] pointer. See the safety documentation of [`create_route`].
-#[allow(rustdoc::private_intra_doc_links)]
+/// # Reference Safety (ref_...)
+/// Any reference data is considered "caller owned".
+///
+/// # Freeing data allocated by the API
+/// To free data returned by the API, use the corresponding `free_...` methods of the API.
+/// It is **not optional** to call them to free data allocated by the API!
 #[no_mangle]
-pub unsafe extern "C" fn command(
-    client_ptr: *const c_void,
-    callback_index: usize,
-    request_type: RequestType,
-    args: *const *mut c_char,
-    arg_count: u32,
-    args_len: *const u32,
-    route_info: *const RouteInfo,
-) {
-    let client = unsafe {
-        // we increment the strong count to ensure that the client is not dropped just because we turned it into an Arc.
-        Arc::increment_strong_count(client_ptr);
-        Arc::from_raw(client_ptr as *mut Client)
+pub extern "C-unwind" fn csharp_create_client_handle(
```

Comment:
all pub extern functions should be marked as `unsafe` and have a safety block in the doc